### PR TITLE
feat: implement new search logic

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [ ] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [ ] contains a reference JIRA issue number like `SQPIT-764`
    - [ ] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

search need to be changed to handle big teams

### Causes (Optional)

there are a lot of changes because thesearch view models were a bit hard to change because of the base search view models

### Solutions

Divide and conquer
1. Reduce the responsibility of of AddMembersToConversationViewModel and NewConverstionViewModel to what there respective names suggest.
2. Add 2 new view models to handle only the search for each Services and Users
3. use the new search use case to search for users 

### Dependencies (Optional)

_If there are some other pull requests related to this one (e.g. new releases of frameworks), specify them here._

Needs releases with:

- [ ] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

#### How to Test

_Briefly describe how this change was tested and if applicable the exact steps taken to verify that it works as expected._

### Notes (Optional)

_Specify here any other facts that you think are important for this issue._

### Attachments (Optional)

_Attachments like images, videos, etc. (drag and drop in the text box)_ 

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [ ] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2496](https://10.10.124.134/job/AR-build-pipeline/2496/) **failed**.

**Comment by AndroidBob** [comment]
Build [2522](https://10.10.124.134/job/AR-build-pipeline/2522/) **failed**.

**Comment by AndroidBob** [comment]
Build [2527](https://10.10.124.134/job/AR-build-pipeline/2527/) **failed**.

**Comment by AndroidBob** [comment]
Build [2537](https://10.10.124.134/job/AR-build-pipeline/2537/) **failed**.

**Comment by AndroidBob** [comment]
Build [2545](https://10.10.124.134/job/AR-build-pipeline/2545/) **failed**.

**Comment by AndroidBob** [comment]
Build [2547](https://10.10.124.134/job/AR-build-pipeline/2547/) **failed**.

**Comment by AndroidBob** [comment]
Build [2549](https://10.10.124.134/job/AR-build-pipeline/2549/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8613678)(8613678)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8613678)(8613678)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [2551](https://10.10.124.134/job/AR-build-pipeline/2551/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8617459)(8617459)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8617459)(8617459)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchAllPeopleScreen.kt [hunk 1]
```diff
@@ -89,51 +101,69 @@ fun SearchAllPeopleScreen(
...
     lazyListState: LazyListState = rememberLazyListState()
 ) {
     val searchPeopleScreenState = rememberSearchPeopleScreenState()
     val context = LocalContext.current
+    val snackbarHostState = LocalSnackbarHostState.current
+    val scope = rememberCoroutineScope()
 
+    val sendConnectionRequest: (UserId) -> Unit = remember {
+        { userId: UserId ->
+            val defrredResult = sendConnectionRequestViewModel.addContact(userId)
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchAllPeopleScreen.kt]:
```suggestion
            val deferredResult = sendConnectionRequestViewModel.addContact(userId)
```

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchAllPeopleScreen.kt [hunk 2]
```diff
@@ -89,51 +101,69 @@ fun SearchAllPeopleScreen(
...
-                            searchQuery = searchQuery,
-                            contactSearchResult = result,
-                            showAllItems = searchPeopleScreenState.publicResultsCollapsed,
-                            onShowAllButtonClicked = searchPeopleScreenState::toggleShowAllPublicResult,
-                            onOpenUserProfile = onOpenUserProfile,
-                            onAddContactClicked = onAddContactClicked
-                        )
-                    }
+            if (contactsSearchResult.isNotEmpty()) {
+                appLogger.d("cccc contactsSearchResult: ${contactsSearchResult.size}")
```
Comment (alexandreferris) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/SearchAllPeopleScreen.kt]: nitpic: Maybe just removing the `cccc` from the logs? Or adding something like:

```suggestion
                appLogger.d("Contacts Search Result Size: ${contactsSearchResult.size}")
```

**Comment by AndroidBob** [comment]
Build [2565](https://10.10.124.134/job/AR-build-pipeline/2565/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8642871)(8642871)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8642871)(8642871)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [2596](https://10.10.124.134/job/AR-build-pipeline/2596/) **failed**.

**Comment by AndroidBob** [comment]
Build [2604](https://10.10.124.134/job/AR-build-pipeline/2604/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8652129)(8652129)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8652129)(8652129)-staging-compat.apk)

**Review by alexandreferris — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [2609](https://10.10.124.134/job/AR-build-pipeline/2609/) **failed**.

**Comment by AndroidBob** [comment]
Build [2613](https://10.10.124.134/job/AR-build-pipeline/2613/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8652620)(8652620)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8652620)(8652620)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/addMemberToConversation/AddMemberToConversationState.kt [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.android.ui.home.conversations.search.addMemberToConversation
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/addMemberToConversation/AddMemberToConversationState.kt]: Not sure what to do, but this package name with camelCase throws me a bit off 😅

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/search/addMemberToConversation/AddMembersToConversationViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.collections.immutable.ImmutableSet
+import kotlinx.collections.immutable.persistentSetOf
+import kotlinx.collections.immutable.toImmutableSet
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import javax.inject.Inject
+
+@HiltViewModel
+class AddMembersToConversationViewModel @Inject constructor(
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/search/addMemberToConversation/AddMembersToConversationViewModel.kt]: I believe this is not tested.

**Review by vitorhugods — COMMENTED** [review]
Looks good.

Just missing tests for a couple of new ViewModels.

Not sure what to do about the camelCasing of package names. It's discouraged, but not forbidden.
I can't think of anything else at the moment, I guess it's fine 🤔

**Review by vitorhugods — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [2649](https://10.10.124.134/job/AR-build-pipeline/2649/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8669310)(8669310)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8669310)(8669310)-staging-compat.apk)

**Pull request merged**